### DISCUSSION TITLE
check_drc - How to use it?

### DISCUSSION THREAD

**Discussion by jcma86** [The-OpenROAD-Project/OpenROAD#4202] [Q&A] [answered]
How can I use "check_drc"?
1. I load a LEF file (`read_lef file.lef`)
2. I load a DEF file (`read_def file.def`) produced after `detailed_route` with still some DRC violations.
3. I run `drt::check_drc -output_file file.rpt`
4. OpenROAD crashes with:
```bash
/usr/include/c++/10/bits/stl_vector.h:1063: std::vector<_Tp, _Alloc>::const_reference std::vector<_Tp, _Alloc>::operator[](std::vector<_Tp, _Alloc>::size_type) const [with _Tp = std::unique_ptr<fr::frAccessPoint>; _Alloc = std::allocator<std::unique_ptr<fr::frAccessPoint> >; std::vector<_Tp, _Alloc>::const_reference = const std::unique_ptr<fr::frAccessPoint>&; std::vector<_Tp, _Alloc>::size_type = long unsigned int]: Assertion '__builtin_expect(__n < this->size(), true)' failed.
Signal 6 received
Stack trace:
 0# 0x000055F65D6A0236 in openroad
 1# 0x00007F3C42497D60 in /lib/x86_64-linux-gnu/libc.so.6
 2# gsignal in /lib/x86_64-linux-gnu/libc.so.6
 3# abort in /lib/x86_64-linux-gnu/libc.so.6
 4# 0x000055F65D6A08A2 in openroad
 5# fr::io::Parser::initRPin_rpin() in openroad
 6# fr::io::Parser::initRPin() in openroad
 7# triton_route::TritonRoute::initGuide() in openroad
 8# triton_route::TritonRoute::checkDRC(char const*, int, int, int, int) in openroad
 9# 0x000055F65DEBB7C8 in openroad
10# TclNRRunCallbacks in /lib/x86_64-linux-gnu/libtcl8.6.so
11# Tcl_RecordAndEvalObj in /lib/x86_64-linux-gnu/libtcl8.6.so
12# Tcl_MainEx in /lib/x86_64-linux-gnu/libtcl8.6.so
13# main in openroad
14# __libc_start_main in /lib/x86_64-linux-gnu/libc.so.6
15# _start in openroad
[1]    4098007 abort      openroad
```

**Comment by maliberty** [comment, marked as answer]
Be aware check_drc is not a general purpose checker and is only verifying the signal routing according to LEF.  It is expecting to see guides as well so you could either
1) Use read_db to load the database that has guides in it
2) Write out the guide and load them later.

1 is probably simpler.

**Reply by jcma86** [reply]
Oh, got it... Thanks for the quick answers!

**Comment by b224hisl** [comment]
After reviewing the source code, I noticed that the `check_drc ` function is only invoked during the detail routing stage. Does this mean that DRC checks are performed exclusively at this stage, and not earlier in the flow? The actual DRC checks are implemented in the `getDRCMarkers`  -> `FlexGCWorker` within TritonRoute.cpp. 

I would like to understand which design rules OpenROAD currently supports for checking. Where can I find a comprehensive list Could you please advise where I should look to investigate this further?

Thank you for your clarification!

**Reply by rovinski** [reply]
DRC is only checked during detailed routing and only for signal nets. PDN has some internal rule checking only during PDN generation.

As for supported DRC rules, it would be easier to list the ones which are _not_ supported. The easiest way is to simply read the LEF file and see if there are any unsupported warnings. Unsupported rules always produce a warning message.

**Reply by b224hisl** [reply]
Thanks for you advice. So, as for the PDN generation rule, which command checks them? I would like to check its source code and see how it do such checking

**Reply by b224hisl** [reply]
And I'd like to know which aspects of the rules KLayout checks—does this overlap with the checks built into OpenROAD itself?

**Comment by maliberty** [comment]
It is not a general purpose drc engine but the engine used by the detailed router.  It will not report violation between fixed objects nor see inside a LEF abstract for detailed checking.  It isn't a replacement for a tool like klayout or magic for gds verification.  There is no official write up for exactly which parts of LEF are supported but we tend to give warnings for unknown constructs.
